enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,616 @@
+* As a MicroShift administrator, I want automated backups of 
+  MicroShift data so in case of a greenboot failure system will rollback
+  and backup will be restored.
+
+### Goals
+
+Goal of the enhancement is to describe implementation roadmap for
+integrating MicroShift with ostree and greenboot in order to provide
+functionality to:
+- Safely update MicroShift version (by backing up the data and
+  restoring it in case of rollback)
+- Migrating internal data (like Kubernetes storage or etcd schema) to
+  newer version
+- Block upgrades in case of version bump being too big
+
+Design aims to implement following principles:
+- Keep it simple, optimize later
+- MicroShift does not own the OS or host
+- MicroShift and all its components are version, upgraded, and rolled back together
+- Be defensive, fail fast
+- Rely on outside intervention as a last resort
+
+### Non-Goals
+
+* Building allowed/blocked version migration graph
+* Handling readiness, and backup and rollback of 3rd party applications
+  (although end user documentation should be provided)
+* Defining updateability for non-ostree systems is left to a future enhancement
+* Protecting against data corruption
+
+## Proposal
+
+### Workflow Description
+
+#### User perspective
+
+**MicroShift administrator** is a human responsible for preparing
+ostree commits and scheduling devices to use these commits.
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots to new commit
+1. Operating System, greenboot, and MicroShift take actions without any additional intervention
+
+#### Implementation perspective
+
+**First boot into commit with MicroShift**
+
+1. First ostree commit is installed
+1. First ostree commit boots
+1. MicroShift starts
+   - No backup action (did not previously run, so there is nothing to backup, and did not previously fail so no restore needed)
+1. MicroShift startup succeeds
+1. All greenboot checks pass
+1. Greenboot runs green scripts
+   - MicroShift green script sets backup mode to "backup"
+
+**Simple host reboot (cont of previous flow)**
+
+1. First ostree commit shuts down
+1. First ostree commit boots
+1. MicroShift starts up
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
+1. MicroShift startup succeeds
+1. All greenboot checks pass
+1. Greenboot runs green scripts
+   - MicroShift green script sets backup mode to "backup" (no change)
+
+**Failed upgrade (cont of previous flow)**
+
+1. Second ostree commit is staged
+1. First ostree commit shuts down
+1. Second ostree commit boots
+1. MicroShift starts up
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
+1. MicroShift startup fails
+1. Greenboot runs red scripts
+    - Set backup mode to "restore"
+1. (Failures may need to repeat to trigger rollback.)
```

Comment:
This needs to be an inherent part of the procedure because we're guaranteed to have more than one attempted boot before the actual rollback occurs.

Reply:
See [this line](https://github.com/openshift/microshift/blob/a45ce756d09e95e9102bbf016bbd201cd37d8e0a/packaging/greenboot/microshift-pre-rollback.sh#L38) on how MicroShift greenboot identifies that case.

Reply:
You mean that word "may" should be removed to indicate that "failure **must** repeat to trigger rollback"?

Reply:
We need to explicitly mention a boot loop in the sequence. It's not obvious to me how the system works correctly when multiple "red" boots are performed until the rollback is attempted by greenboot.